File: hackerrank/week_of_code/33/transform_to_palindrome.py

from collections import defaultdict



trans = defaultdict(list)
s = ""
pool = {}
memo = {}


def ff(n):
    t = 0
    visited = set()

    for i in range(1, n+1):
        if i not in visited:
            st = [i]
            visited.add(i)
            while st:
                c = st.pop()
                pool[c] = t

                for neigh in trans[c]:
                    if neigh not in visited:
                        visited.add(neigh)
                        st.append(neigh)

            t += 1



def can_equate(a, b):
    return pool[a] == pool[b]



def transform(first, last):
    if first >= last:
        return 0
    elif (first, last) in memo:
        return memo[(first, last)]
    else:
        if first == last or can_equate(s[first], s[last]):
            r = transform(first + 1, last - 1)
        else:
            r = 1 + min(transform(first + 1, last), transform(first, last - 1))

        memo[(first, last)] = r
        return r



def main():
    global s

    n, k, m = [int(x) for x in raw_input().split()]

    for _ in range(k):
        a, b = [int(x) for x in raw_input().split()]
        trans[b].append(a)
        trans[a].append(b)

    ff(n)

    s = [int(x) for x in raw_input().split()]

    print len(s) - transform(0, m - 1)


main()

File: hackerrank/week_of_code/33/transform_to_palindrome.cpp

#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <stack>
#include <cstring>
#include <algorithm>
using namespace std;

int pool[100001];
int memo[1001][1001];
unordered_map<int, vector<int>> trans;
vector<int> s;
int visited[100001];


void flood_fill(int n) {
    int t = 0;

    for (int i = 1; i <= n; ++i) {
        if (!visited[i]) {
            stack<int> st;
            st.push(i);
            visited[i] = 1;
            int cur;

            while (!st.empty()) {
                cur = st.top();
                st.pop();
                pool[cur] = t;

                for (int j = 0; j < trans[cur].size(); ++j)
                    if (!visited[trans[cur][j]]) {
                        visited[trans[cur][j]] = 1;
                        st.push(trans[cur][j]);
                    }
            }

            t += 1;
        }
    }
}


int transform(int first, int last) {
    if (first >= last)
        return 0;
    else if (memo[first][last] != -1)
        return memo[first][last];
    else {
        if (pool[s[first]] == pool[s[last]])
            return memo[first][last] = transform(first + 1, last - 1);
        else
            return memo[first][last] = 1 + min(transform(first + 1, last), transform(first, last - 1));
    }
}

int main() {
    int n, k, m, a, b;
    cin >> n >> k >> m;

    for (int i = 0; i < k; ++i) {
        cin >> a >> b;
        trans[a].push_back(b);
        trans[b].push_back(a);
    }

    flood_fill(n);

    for (int i = 0; i < m; ++i) {
        cin >> a;
        s.push_back(a);
    }

    memset(memo, -1, sizeof(memo));
    cout << m - transform(0, m-1) << endl;

    return 0;
}
